### PR TITLE
Very minimal machine state and service implementation for dqlite

### PR DESCRIPTION
**Pull request by wallyworld**
Add very minimal dqlite persistence support for machines. This is to provide a parent record which block devices can relate to. For now, just the machine id and life (alive) is saved; the only methods on the service are `Save` and `Delete`. A follow up PR will wire this through so that when a machine is created or removed, the dqlite records are updated also.

## QA steps

just unit tests at this stage

## Links

**Jira card:** JUJU-5219

### CODE REVIEW

File: domain/machine/state/state_test.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(machineCount, gc.Equals, 0)
+}
+
+func (s *stateSuite) insertBlockDevice(c *gc.C, bd blockdevice.BlockDevice, blockDeviceUUID, machineId string) {
```
Comment (wallyworld) [inline domain/machine/state/state_test.go:109]: lifted from existing block device unit test

File: domain/machine/state/state.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	s.logger.Debugf("upserted machine %q", machineId)
+
+	return nil
+}
+
+func upsertMachine(ctx context.Context, tx *sqlair.TX, machineId string) error {
```
Comment (manadart) [inline domain/machine/state/state.go:58]: I think extracting this for re-use as a bootstrap concern is taking DRY too far.

To achieve it, we're pushing a slew of logic into a write transaction (statement preparations for example), but at bootstrap time we know there won't be any conflicts, as the machine can't exist - it's never an upsert.

I'd rather duplicate the minimum required for a bootstrap machine creation, and keep the more frequent case lean.

Comment (wallyworld) [reply domain/machine/state/state.go]: Refactored it

File: domain/machine/bootstrap/bootstrap.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain/life"
+)
+
+// InsertBootstrapMachine inserts the initial machine during bootstrap.
+// TODO - this just creates a minimal row for now.
+func InsertBootstrapMachine(machineId string) func(context.Context, database.TxnRunner) error {
```
Comment (tlm) [inline domain/machine/bootstrap/bootstrap.go:19]: Can we remove the word Bootstrap? Sort of implied based on the package.

File: domain/machine/state/state.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+			return errors.Annotatef(err, "deleting net node for machine  %q", machineId)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	s.logger.Debugf("deleted machine %q", machineId)
```
Comment (tlm) [inline domain/machine/state/state.go:171]: Does this really need to exist and if so it should be coming from above the services layer. The fact the stack hasn't errored is the exact same meaning.

File: domain/machine/state/state.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+		if err := tx.Query(ctx, createMachineStmt, createParams).Run(); err != nil {
+			return errors.Annotatef(err, "creating machine row for machine %q", machineId)
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	s.logger.Debugf("upserted machine %q", machineId)
```
Comment (tlm) [inline domain/machine/state/state.go:106]: Same as below. Is this really needed? What is the value this is adding?

File: domain/machine/service/service.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+func NewService(st State) *Service {
+	return &Service{
+		st: st,
+	}
+}
+
+// Save inserts or updates the specified machine.
+func (s *Service) Save(ctx context.Context, machineId string) error {
+	err := s.st.UpsertMachine(ctx, machineId)
+	return errors.Annotatef(err, "updating machine %q", machineId)
```
Comment (tlm) [inline domain/machine/service/service.go:36]: s/updating/saving

File: domain/machine/service/service_test.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	s.state.EXPECT().UpsertMachine(gomock.Any(), "666").Return(nil)
+
+	err := NewService(s.state).Save(context.Background(), "666")
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *serviceSuite) TestUpdateError(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.state.EXPECT().UpsertMachine(gomock.Any(), "666").Return(errors.New("boom"))
```
Comment (tlm) [inline domain/machine/service/service_test.go:41]: Can you update this to:

```
rErr := errors.New("boom")
s.state.EXPECT().UpsertMachine(gomock.Any(), "666").Return(rErr)
err := NewService(s.state).Save(context.Background(), "666")
c.Check(err, jc.ErrorIs, rErr)
c.Check(err, gc.ErrorMatches, `updating machine "666": boom`)
```

File: domain/machine/service/service_test.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *serviceSuite) TestDeleteError(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.state.EXPECT().DeleteMachine(gomock.Any(), "666").Return(errors.New("boom"))
+
+	err := NewService(s.state).Delete(context.Background(), "666")
+	c.Assert(err, gc.ErrorMatches, `deleting machine "666": boom`)
```
Comment (tlm) [inline domain/machine/service/service_test.go:62]: Same comment as above. Better to check the error type and not the string. By checking the string we are saying this exact string is our interface.

**Review by tlm — APPROVED** [review]
Would like the PR feedback to be applied. Conditional approval.

File: domain/machine/bootstrap/bootstrap.go (hunk 3 above)
Comment (wallyworld) [reply domain/machine/bootstrap/bootstrap.go:19]: Totally understand the concern, but here it literally is called the "bootstrap machine". I could maybe rename the method to InsertMachineZero (which is what it always will be) and lose the machine id param.

File: domain/machine/state/state.go (hunk 4 above)
Comment (wallyworld) [reply domain/machine/state/state.go:171]: Yeah, was being consistent with what was done elsewhere. Can remove.

File: domain/machine/bootstrap/bootstrap.go (hunk 3 above)
Comment (wallyworld) [reply domain/machine/bootstrap/bootstrap.go:19]: Went with InsertMachine() for now; can revisit later when it's all fully implemented.

File: domain/machine/service/service.go (hunk 6 above)
Comment (wallyworld) [reply domain/machine/service/service.go:36]: yup

File: domain/machine/service/service_test.go (hunk 8 above)
Comment (wallyworld) [reply domain/machine/service/service_test.go:62]: We typically use regexps in the `gc.ErrorMatches` just to check the core message we care about; the fluff around it is typically unimportant. Here the message was so simple a regexp was not needed. However, I'll make these requested changes here but we will have 10000000000000 other places to fix if we want to be consistent.

Comment (wallyworld) [reply domain/machine/service/service_test.go:62]: Also, our test have very deliberately tested (part of) the error string as that forms part of the UX experience and we've caught various usability issues doing this.

**Pull request merged**